### PR TITLE
Python 3.12 support

### PR DESCRIPTION
**Pull request by johnnynunez**
#13072
- Python 3.8 deprecation (https://devguide.python.org/versions/)
image

### CODE REVIEW

**Comment by travishathaway** [comment]
@johnnynunez,

We're expecting 3.12 to be ready on defaults by the end of this week.

**Comment by mbargull** [comment]
> We're expecting 3.12 to be ready on defaults by the end of this week.

That'd be a good pitch to poke someone at Anaconda to take a look at https://github.com/Anaconda-Platform/anaconda-client/pull/691 , right ;).

**Comment by jezdez** [comment]
This depends on https://github.com/conda/conda/pull/13142

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -70,7 +70,7 @@ jobs:
       matrix:
         # test lower version (w/ defaults) and upper version (w/ defaults and conda-forge)
         default-channel: ['defaults', 'conda-forge']
-        python-version: ['3.8', '3.11']
+        python-version: ['3.9', '3.12']
```
Comment (jezdez) [inline .github/workflows/tests.yml:73]:
```suggestion
        python-version: ['3.8', '3.12']
```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -156,16 +156,16 @@ jobs:
       matrix:
         # test all lower versions (w/ defaults) and upper version (w/ defaults and conda-forge)
         default-channel: ['defaults', 'conda-forge']
-        python-version: ['3.8', '3.9', '3.10', '3.11']
+        python-version: ['3.9', '3.10', '3.11', '3.12']
```
Comment (jezdez) [inline .github/workflows/tests.yml:159]:
```suggestion
        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
```

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -156,16 +156,16 @@ jobs:
...
         test-type: ['unit', 'integration']
         test-group: ['1', '2', '3']
         exclude:
-          - default-channel: 'conda-forge'
-            python-version: '3.8'
           - default-channel: 'conda-forge'
```
Comment (jezdez) [inline .github/workflows/tests.yml:163]:
```suggestion
          - default-channel: 'conda-forge'
            python-version: '3.8'
          - default-channel: 'conda-forge'
```

File: docs/source/user-guide/tasks/manage-python.rst [hunk 4]
```diff
@@ -10,7 +10,7 @@ Managing Python
 Conda treats Python the same as any other package, so it is easy
 to manage and update multiple installations.
 
-Conda supports Python 3.8, 3.9, 3.10, and 3.11.
+Conda supports Python 3.9, 3.10, 3.11 and 3.12.
```
Comment (jezdez) [inline docs/source/user-guide/tasks/manage-python.rst:13]:
```suggestion
Conda supports Python 3.8, 3.9, 3.10, 3.11 and 3.12.
```

File: recipe/conda_build_config.yaml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 python:
-  - 3.8
   - 3.9
```
Comment (jezdez) [inline recipe/conda_build_config.yaml:2]:
```suggestion
  - 3.8
  - 3.9
```

**Review by jezdez — CHANGES_REQUESTED** [review]
Just because we're adding Python 3.12, doesn't mean we're removing 3.8.

**Review by jezdez — APPROVED** [review]
As soon as #13142 is merged we can do this.

**Comment by johnnynunez** [comment]
> @johnnynunez,
> 
> We're expecting 3.12 to be ready on defaults by the end of this week.

3.11.6?

**Comment by dbast** [comment]
What about extending also https://github.com/conda/conda/blob/main/pyproject.toml#L20 ?

**Comment by kenodegard** [comment]
hm looks like menuinst doesn't support Python 3.12 yet: https://github.com/conda/menuinst/pull/164

File: conda/gateways/repodata/__init__.py [hunk 6]
```diff
@@ -456,19 +456,19 @@ def should_check_format(self, format: str) -> bool:
...
         return super().__setitem__(key, item)
 
-    def __missing__(self, key: str):
-        if key in self._aliased:
-            key = self._aliased[key]
-        else:
-            raise KeyError(key)
+    def __getitem__(self, key: str) -> Any:
+        key = self._aliased.get(key, key)
         return super().__getitem__(key)
```
Comment (kenodegard) [inline conda/gateways/repodata/__init__.py:472]: cc @dholth

**Review by kenodegard — APPROVED** [review]

File: conda/gateways/repodata/__init__.py (hunk 6 above)
Comment (jaimergp) [reply conda/gateways/repodata/__init__.py:472]: Is this needed as part of the 3.12 support? 🤔

Comment (dholth) [reply conda/gateways/repodata/__init__.py:472]: Yes, 3.12 changed the way UserDict works.

Comment (dholth) [reply conda/gateways/repodata/__init__.py:472]: It would be a good idea to link to the issue # in a comment.

Comment (kenodegard) [reply conda/gateways/repodata/__init__.py:472]: See https://github.com/python/cpython/pull/17910

**Pull request merged**